### DISCUSSION TITLE
Lesson 14's vm.wrap & vm.roll.

### DISCUSSION THREAD

**Discussion by Glow-in-the-dark** [Cyfrin/foundry-full-course-cu#574] [Q&A] [answered]
In the example "MyGovernorTest.t.sol".
![Screenshot 2023-08-07 at 11 01 11 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/2751458/2b61b9a4-490b-455d-a66e-f7aeb844023a)

2 questions...

**1st Question:**
1) When i do a console.log, on block.number and block.timestamp, why is block.timestamp showing the same as block.number?
![Screenshot 2023-08-07 at 11 23 43 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/2751458/f17cd4b0-3e7c-44a4-b293-eaf80f4f4d62)
shouldn't block.timestamp gives the unix timestamp of the block that is included in the chain?
<img width="500" alt="Screenshot 2023-08-07 at 11 45 34 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/2751458/88955ff2-985d-4d4b-991f-bf42c6db8565">

- block.timestamp: is a property that represents the time when the block was mined. It is set by the miner of the block, based on their local clock. It is expressed in Unix time?
- block.number is more of the location of which the block.

--------------------------------------------------------------------

**2nd Question**
2) Why do we use both `vm.warp` && `vm.roll` ?
Understand that warp is used for "block.timestamp", and roll is used for "block.number"
Why wouldn't just one enough? for example, just using vm.roll, and not vm.warp, and vice versa ?
Any good counter examples which helps to explain the difference better ?

**Comment by theirrationalone** [comment, marked as answer]
# Answer 1:

We often do console.log to see the content of variables & constants, console.log prints the value of the variable Or constant on to the terminal. 

And It's just a coincidence that you got identical values of timestamp And block number. 

Where timestamp value is the time in time unit And, 
Block number is a count of total number of blocks have been mined upto now. 

# Answer 2:

# vm.warp:

We use vm.warp for time travelling. 
Let's Say for example:...

at Very first keep this in mind ->, When i deployed the contract "raffle" also called as lottery, It was 9:30 AM,

So we have a timestamp of 90:30AM <- called as lastTimestamp, and
We choose an interval of = 30 mintues,
and then i performed entry of new player into lottery using that modifier... "raffleEnteredAndTimePassed", Okay...

And We have tools/facilities to do things i.e.,

warp for time-travelling.

roll for mining blocks.

Thanks to foundry again.

As you know, The modifier "raffleEnteredAndTimePassed" also performs warp & roll after putting new player into lottery/raffle,
So when i performed or put new player into lottery, warp & roll might also happened,
And please ignore roll here,
So Warp affected the time and that lastTimestamp updated to 10:00 AM + 1.

Now we have...

Players = 1 <- into lottery/raffle.
lastTimestamp = 10:00 AM + 1,



# vm.roll:

We use vm.roll to rollup Or mine the block in which our transaction resides, that means our transaction is mined & validated and successfully added into the blockchain so in future we will have a proof of our work. 

For more [click here](https://github.com/Cyfrin/foundry-full-course-f23/discussions/468#discussioncomment-6637107)

**Reply by Glow-in-the-dark** [reply]
ahhh thanks so much @theirrationalone. Really helpful!!

So vm.warp ... warps us to that specific time. ( tx can be there, but not necessarily the blocks are mined )
so by using vm.roll(block.number) ... it mines all the transactions up till that specific block.number ?

**Comment by theirrationalone** [comment]
> ahhh thanks so much @theirrationalone. Really helpful!!
> 
> So vm.wrap ... wraps us to that specific time. ( tx can be there, but not necessarily the blocks are mined ) so by using vm.roll(block.number) ... it mines all the transactions up till that specific block.number ?

Thunder ⚡⚡⚡ You Got Enough!

**Comment by MarcoW3** [comment]
Answering to the first question, I believe it depends on the fact we are using the same addend for time-changing and block-chaning. VOTING_PERIOD is about blocks, and not about time. It's not obvious that the number of blocks added is equal to a week-time.
